openshift-knative/serverless-operator
# Detached image override

With this PR, the image override is implemented for Kafka components.

To help you follow what is going on:

- 0f78480: Just copy&paste the image override mechanism from upstream **as is**. It would be nice to use it as a library, but it is not possible because `KnativeKafka` doesn't support overrides with CR spec.
- 155b07d: Modify the *as is* image override mechanism to get rid of default registry and image cache. Also use the transform in the controller
- 3f2f2f6: Make the necessary changes in the CSV with actual image overrides. Naming clash in `DISPATCHER_IMAGE` env var override.
- ad558aa: Add scopes to image overrides. "DISPATCHER_IMAGE" is now "KAFKA_DISPATCHER_IMAGE" and "EVENTING_DISPATCHER_IMAGE".

About "scoping" image overrides:
- There is an env var which we should override: `DISPATCHER_IMAGE` of `kafka-ch-controller` deployment. However, we already have an override for that env var in Knative Eventing.
- The copied upstream image override mechanism doesn't check the deployment or container name when it is done with env vars.
- Thus, I created "scopes" for image overrides

Chat with Markus: https://coreos.slack.com/archives/CD87JDUB0/p1601892394404600

The scoping should later be used in serving and eventing. That way, we wouldn't have e.g. eventing related image overrides in serving CR when we mutate the CR.

Verification instructions:

1. Enable the controller by uncommenting the code in `webhook/add_knativekafka.go` and `controller/add_knativekafka.go`

2. `make images` && `make install` --> deploys S-O and installs KnativeServing and KnativeEventing

3. Need to manually create the CRD since it is not in the CSV yet
```
kubectl apply -f knative-operator/deploy/crds/operator_v1alpha1_knativekafka_crd.yaml
```

4. Install KnativeKafka:
```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.serverless.openshift.io/v1alpha1
kind: KnativeKafka
metadata:
  name: example
  namespace: knative-eventing
spec:
  source:
    enabled: true
  channel:
    enabled: true
    bootstrapServers: foo.example.com:1234
...
EOS
```

5. Watch the status of `KnativeKafka` CR, should be fine.
```
kubectl get knativekafka -n knative-eventing -o json | jq -r '.items[] | .metadata.name + ":" + (.status.conditions[] | .type + ":" + .status)'
```

6. Check the images for deployments in `knative-eventing`, `knative-serving`, `knative-sources` and `knative-serving-ingress` namespaces. No gcr.io should be there

7. Clean up:
```
kubectl delete knativekafkas example -n knative-eventing
```
